### PR TITLE
*: add support for comments in node/entities

### LINKED ISSUE #3: gatherer: load comments best approach
Hi guys,

First of all thanks for your contribution to the open source community with such a great tool, well done.

I would like to take the task "Load comments from proto files into gathered AST for easy access" but I have some doubts about what would be the best approach. Basically the question is, what do you prefer consume memory or CPU? I mean we can tackle this by traversing `File`s [`SourceCodeInfo_Location`](https://github.com/golang/protobuf/blob/master/protoc-gen-go/descriptor/descriptor.pb.go#L1821) each time an entity is being processed, or store a lookup map at the very beginning.

WDYT?

### PR DESCRIPTION
**Pull request by glerchundi**
Closes #3

### CODE REVIEW

File: gatherer.go [hunk 1]
```diff
@@ -421,4 +452,90 @@ func (g *gatherer) resolveLookupName(e Entity) string {
...
+	// an index number thus always leading to even paths.
+	if len(path)%2 != 0 {
+		return "", errors.New("path must have even elements")
+	}
+
+	// tail call
+	var fn func(interface {
+		GetName() string
+	}, []int32, *[]string) error
+	fn = func(parent interface {
```
Comment (rodaine) [inline gatherer.go]: just `fn :=` should be fine, no?

File: gatherer.go [hunk 2]
```diff
@@ -421,4 +452,90 @@ func (g *gatherer) resolveLookupName(e Entity) string {
 	return e.lookupName()
 }
 
+func (g *gatherer) nameByPath(f *descriptor.FileDescriptorProto, path []int32) (string, error) {
```
Comment (rodaine) [inline gatherer.go]: Can we get some tests on this?

File: file.go [hunk 3]
```diff
@@ -145,3 +149,10 @@ func (f *file) addService(s Service) {
 }
 
 func (f *file) addMapEntry(m Message) { panic("cannot add map entry directly to file") }
+
+func (f *file) lookupComments(name string) string {
+	if c, ok := f.comments[name]; ok {
+		return c
+	}
+	return ""
```
Comment (rodaine) [inline file.go]: this can probably just be simplified to `return f.comments[name]` no?

Comment (glerchundi) [reply file.go]: I think so, will do.

File: gatherer.go [hunk 4]
```diff
@@ -47,13 +62,14 @@ func (g *gatherer) hydratePackage(f *generator.FileDescriptor) Package {
 		fd:         f,
 		name:       name,
 		importPath: goImportPath(g.Generator.Unwrap(), f),
+		comments:   comments[fmt.Sprintf(".%s", name)],
```
[the remark this answers is not in the document]
Comment (glerchundi) [reply gatherer.go]: Ok, let me check what can I do.

File: gatherer.go (hunk 2 above)
Comment (glerchundi) [reply gatherer.go]: Ok.

File: gatherer.go (hunk 1 above)
Comment (glerchundi) [reply gatherer.go]: No, because it's used inside the functor itself and will not be discoverable.

**Comment by glerchundi** [comment]
> Thanks for the PR! This has been a highly requested feature :)

It was a pleasure, thanks to you all for open sourcing this high quality piece of code.

> Just a few nits, and please add some tests to cover your changes, particularly around nameByPath. Also, update the readme to reflect this feature is available.

No problem, will do. I tested it manually by commenting every entity in the kitchen example, like:

```
[...]
// .kitchen.Style enum
enum Style {
    // .kitchen.Style.CONTEMPORARY enum value
    CONTEMPORARY  = 0;

    // .kitchen.Style.COTTAGE_CHARM enum value
    COTTAGE_CHARM = 1;

    // .kitchen.Style.MEDITERRANEAN enum value
    MEDITERRANEAN = 2;

    // .kitchen.Style.COASTAL enum value
    COASTAL       = 3;

    // .kitchen.Style.OLD_WORLD enum value
    OLD_WORLD     = 4;

    // .kitchen.Style.TRADITIONAL enum value
    TRADITIONAL   = 5;
}
[...]
```

Do you want me to upstream those changes too?

I will come back to you once I make the changes.

**Comment by glerchundi** [comment]
@rodaine i think it's ready for the second pass, will squash commits once we decide it's ok.

**Comment by rodaine** [comment]
Will take a look in a bit! Don't worry about squashing (it'll be squashed on merge automatically)

File: gatherer_test.go [hunk 5]
```diff
@@ -695,6 +697,120 @@ func TestGatherer_SeenObj(t *testing.T) {
 	assert.Equal(t, m, e)
 }
 
+func TestGatherer_NameByPath(t *testing.T) {
```
Comment (rodaine) [inline gatherer_test.go]: can we add a `t.Parallel()` to this as well as in the subtests below?

Comment (glerchundi) [reply gatherer_test.go]: of course.

**Comment by glerchundi** [comment]
@rodaine i think it's ready for the last pass, i slightly modified the code so that:
* Regarding to comments initialization/modification: I avoided the type check in runtime by adding a new method to the `Package` interface called `setComments`.
* Instead of using the first non-empty comments in a file, I merged all of them so that we don't give any precedence to any file.

**Comment by glerchundi** [comment]
You’re welcome!!

**Pull request merged**